libsolidity/lsp/ReferenceCollector.h

```diff
@@ -0,0 +1,55 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#pragma once
+
+#include <libsolidity/ast/ASTForward.h>
+#include <libsolidity/ast/ASTVisitor.h>
+#include <liblsp/Server.h>
+
+namespace solidity::lsp
+{
+
+class ReferenceCollector: public frontend::ASTConstVisitor
+{
+private:
+	frontend::Declaration const& m_declaration;
+	std::string const& m_sourceIdentifierName;
+	std::vector<::lsp::DocumentHighlight> m_result;
+
+public:
+	ReferenceCollector(frontend::Declaration const& _declaration, std::string const& _sourceIdentifierName);
+
+	std::vector<::lsp::DocumentHighlight> take() { return std::move(m_result); }
```

Comment:
After this is called, the result is in an unspecified state. I don't think it is important to avoid this simple copy.

Reply:
Does it need to be public at all? Wouldn't it be enough to have a single public static function?

Reply:
actually I think I agree. I'll change this.